lnurl-allowance.md

```diff
@@ -0,0 +1,197 @@
+# LNURL-allowance
+
+The LNURL Allowance scheme allows a `LN SERVICE` and a `LN WALLET` to establish a `pairing` over a WebSocket connection. The idea is that `LN SERVICE` will provide an LNURL that returns some parameters (including a socket URL) to `LN WALLET`. `LN WALLET` will diplay a pairing dialog and once accepted by the user, `LN WALLET` will start a socket connection to `LN SERVICE`. This will allow for `LN WALLET` to pay/receive funds to/from `LN SERVICE` without user interaction or QR Code scanning, up to the established allowance amount.
+
+## Scan QR code to establish allowance/pairing
+
+### WALLET to SERVICE interaction flow:
+
+1. User scans a LNURL QR code or accesses a `lightning:LNURL..` link with `LN WALLET` and `LN WALLET` decodes LNURL.
+
+2. `LN WALLET` makes a GET request to `LN SERVICE` using the decoded LNURL.
+
+3. `LN WALLET` gets JSON response from `LN SERVICE` of form:
+
+```
+{
+    socket: String, // the URL from LN SERVICE which will accept the WebSocket connections from the WALLET
+    k1: String, // random or non-random string to identify the WALLET when receiving connection to socket
+    recommendedAllowanceAmount: Millsatoshis, // amount of satoshis the SERVICE recommends WALLET to ask user for this allowance/pairing
```

Comment:
could this be changed to `minAllowanceAmount` and `maxAllowanceAmount`, like with LNURL-withdraw?

Reply:
With the setup we put forward, the WALLET is constantly telling the SERVICE their `balance` through the `balance` message (which also serves as a PING). The reason we don't set a `maximum` amount is that if the WALLET runs out of funds, we'd need to have another message type to `requestNewAllowance` sort of thing. If the WALLET/user would like to `increase` their balance, we can let the WALLET handle that, and provide easy-to-use buttons/UI to handle the `add 5000 sats to balance` sort of thing. The SERVICE just needs to receive that value and know whether the balance is enough for given `interaction/gameplay/whatever`. By setting an upper bound limit, we also increase the amount of interaction the user would have to go through to `increase it`.

The idea behind a `recommendedAllowanceAmount` is that the SERVICE can let the WALLET know what they recommend: e.g. in a game scenario, let's say you usually/likely need 1000 sats to complete the given mission, therefore the SERVICE can suggest/recommend 1000 sats, and you as the WALLET owner can complete given mission before having to `increase balance`. If we don't set a `minimum/recommended` value, the WALLET could default to 50 sats for example, and then the user has to stop and increase their balance right away. The goal for this spec is to minimize the burden around user interactions for little microtransactions like this.

I'm not against the idea of recommending a lower bound and an upper bound, but it should be a recommendation, not a hard limit `maxAllowanceAmount` if that makes sense?

Reply:
Thanks for explaining

Reply:
Now I'm thinking that since it is just a suggestion to the wallet maybe it could have a `minAllowanceAmount`/`maxAllowanceAmount` just to be similar to the other protocols, but this is a minor thing.

Reply:
I'm not against having an upper and a lower bound defined by the SERVICE. Some questions I have:

1 - Can the WALLET put an allowance value `larger than` the `maxAllowanceAmount`?
2 - Can the WALLET put an allowance value `smaller than` the `minAllowanceAmount`?

I understand these are recommendations, but should we have them be more than that? Such that the WALLET has to perform the allowance at the `very least` on the `minAllowanceAmount` and at the `very most` the `maxAllowanceAmount`? Given the WALLET is the one that handles the actual management of the funds in the `allowance`, this shouldn't be an issue either way.

Reply:
Yes, I think the wallet SHOULD enforce that on the user, but it remains the decision of the wallet + user, in whatever decision-making process these two agree on.